gef-ui/server/controllers/unissued-facilities/facility-end-date/index.ts

```diff
@@ -0,0 +1,217 @@
+
+  if (!isFacilityEndDateEnabledOnGefVersion(parseDealVersion(deal.version)) || !facility.isUsingFacilityEndDate) {
+    return null;
+  }
+
+  const facilityEndDateViewModel: FacilityEndDateViewModel = {
+    dealId,
+    facilityId,
+    previousPage,
+  };
+
+  if ('facilityEndDate' in facility && typeof facility.facilityEndDate === 'string') {
+    const facilityEndDate = parseISO(facility.facilityEndDate);
+    facilityEndDateViewModel.facilityEndDate = {
+      day: String(getDate(facilityEndDate)),
+      month: String(getMonth(facilityEndDate) + 1),
+      year: String(getYear(facilityEndDate)),
+    };
+  }
+
+  return facilityEndDateViewModel;
+};
+
+export const getFacilityEndDateFromUnissuedFacilitiesPage = async (req: GetFacilityEndDateRequest, res: Response) => {
+  const {
+    params: { dealId, facilityId },
+  } = req;
+
+  const previousPage = `/gef/application-details/${dealId}/unissued-facilities/${facilityId}/about`;
+
+  try {
+    const facilityEndDateViewModel = await getFacilityEndDateViewModel(req, previousPage);
+
+    if (!facilityEndDateViewModel) {
+      return res.redirect(previousPage);
+    }
+
+    return res.render('partials/facility-end-date.njk', facilityEndDateViewModel);
+  } catch (error) {
+    console.error(error);
+    return res.render('partials/problem-with-service.njk');
+  }
+};
+
+export const getFacilityEndDateFromApplicationPreviewPage = async (req: GetFacilityEndDateRequest, res: Response) => {
+  const {
+    params: { dealId, facilityId },
+  } = req;
+
+  const previousPage = `/gef/application-details/${dealId}/unissued-facilities/${facilityId}/change`;
+
+  try {
+    const facilityEndDateViewModel = await getFacilityEndDateViewModel(req, previousPage);
+
+    if (!facilityEndDateViewModel) {
+      return res.redirect(previousPage);
+    }
+    return res.render('partials/facility-end-date.njk', facilityEndDateViewModel);
+  } catch (error) {
+    console.error(error);
+    return res.render('partials/problem-with-service.njk');
+  }
+};
+
+const getCoverStartDateOrStartOfToday = (facility: Record<string, unknown>): Date => {
+  if (typeof facility.coverStartDate === 'string') {
+    return startOfDay(parseISO(facility.coverStartDate));
+  }
+
+  if (!facility.coverStartDate) {
+    return startOfDay(new Date());
+  }
+
+  throw new Error('Invalid coverStartDate');
+};
+
+export const validateAndReturnErrorsOrUpdateFacilityEndDate = async (req: PostFacilityEndDateRequest): Promise<ValidationError[] | null> => {
+  const {
+    params: { dealId, facilityId },
+    body: { 'facility-end-date-year': facilityEndDateYear, 'facility-end-date-month': facilityEndDateMonth, 'facility-end-date-day': facilityEndDateDay },
+    query: { saveAndReturn },
+  } = req;
+
+  const { userToken, user } = asLoggedInUserSession(req.session);
+
+  const facilityEndDateIsBlank = !facilityEndDateYear && !facilityEndDateMonth && !facilityEndDateDay;
+
+  if (isTrueSet(saveAndReturn) && facilityEndDateIsBlank) {
```

Comment:
Things like this, is it worth commenting something along the lines of 
```
If the user has deleted the facility end date and saves and returns to details, we do not save the change in facility end date to blank
```
? Just to show that this behaviour is intentional